recipe/meta.yaml

```diff
@@ -12,20 +12,20 @@ source:
   sha256: {{ sha256 }}
 
 build:
-  number: 0
+  number: 1
   skip: true  # [win]
 
 requirements:
   build:
-    - openblas
+    - openblas 0.2.20|0.2.20.*
```

Comment:
Can you try to remove the pins in the `meta.yaml` and re-render again?
Latest conda-smithy should be able to apply the pins automatically via the `ci_support/<platform>_.yaml` files.

Reply:
@jakirkham is mostly correct in https://github.com/conda-forge/staged-recipes/pull/5726#discussion_r184173719

I'd advise an alternate pattern.  In the build deps, remove all pins.  Conda-build 3 will do the right thing here.

In the run deps, you won't get automatic pinning until the upstreams are edited to include run_exports and rebuilt.  So, for those, you should do:

``` 
- {{ pin_compatible('openblas', max_pin='x.x.x') }}
- {{ pin_compatible('hdf5', max_pin='x.x.x') }}
```

These can be completely deleted in run deps once the upstreams have run_exports.

Reply:
Note that the number of x's is the number of places to pin.  Since both hdf5 and openblas pin 3 places in @jakirkham's request, I've reflected that in my suggestion here.

Reply:
> We still need `run_exports` on the dependencies for this to work properly 100%, but I wonder that would `ci_support/<platform>_.yaml` if you drop the pins here.

As the package was already built unpinned, we already know this won't work the way we intend it to until the dependencies are rebuilt with `run_exports`.

> In the run deps, you won't get automatic pinning until the upstreams are edited to include run_exports and rebuilt.  So, for those, you should do:
>
> ``` 
> - {{ pin_compatible('openblas', max_pin='x.x.x') }}
> - {{ pin_compatible('hdf5', max_pin='x.x.x') }}
> ```
>
> These can be completely deleted in run deps once the upstreams have run_exports.

Thank @msarahan. Would much prefer this to manually entering these values in `meta.yaml`. So let's go for it. Happy to recommend this strategy to people until we have rebuilt upstream dependencies with `run_exports`.

cc @conda-forge/core @conda-forge/staged-recipes (for awareness)

Reply:
just be careful with the number of x's.  It is not random!  Make sure anyone reviewing the recipes matches up the number of x's with the number of places pinned in our conda-forge-pinning config yaml file.

Reply:
Isn't this the same as https://github.com/conda-forge/conda-forge-pinning-feedstock/issues/4? The reason not to use it was because the pinning info should be in the recipe of the pinning package, not `conda-forge-pinning` or downstream pacakges like this one.

Reply:
That is a bad approach generally because the compatibility of the upstream package can change with its version.  It really should be done with run_exports, where you can change run_exports for a given recipe revision that will affect only that version of a package.

Until it is done with run_exports, your approach there, or my proposed approach would both work.

Reply:
I agree with the first part. Until it is done with run_exports, I'm not a fan of your approach because we have to maintain this info in each downstream recipe and as you said we have to be careful with the number of x's in each recipe vs being careful in just conda-forge-pinning.

Reply:
PS: either way, you'll want to edit the recipe to remove the run dependencies after run_exports are adding upstream.  If you don't remove the run deps with your approach, your compatibility bounds may differ from the run_exports setting.  This will give you more bounds.  Conda will use the tightest merged constraint from all of the bounds provided.

For example, if you pinned to 'x.x.x', but the upstream author knows that 'x.x' is sufficient and puts that in run_exports, your global 'x.x.x' takes priority because it is more restrictive, and all of your packages will be pinned too tightly.

Reply:
> either way, you'll want to edit the recipe to remove the run dependencies after run_exports are adding upstream.

I guess that there is no escaping this and we should focus our efforts on re-build the recipes with `run_exports` ASAP.

Reply:
OK, I'm fine with doing it in conda-forge-pinning instead and advising people to remove pins from the build and run deps.

You'll just put the dep name in build and run.

I'm still not OK with doing this for boost or numpy, for the record.  I'm fine with it as an interim solution for packages that we'll eventually have run_exports for.

For numpy and boost, I think @mingwandroid was correct when he recommended that we have dedicated libboost/numpy-(dev something) that have run_exports, and it is then not ambiguous how the packages are being used.

Reply:
Yeah, this sounds like less work in the long run. Thanks for mentioning this alternative, @isuruf.

What's the issue with `boost`?

Is the issue with `numpy` just the pure Python packages that depend on it? Honestly I don't think that is as much of a problem as people would imagine. Most pure Python packages that use `numpy` are adding it to `run` only (not `build` soon to be `host`). If `conda-build` is still pinning things that are `run` only dependency, we should discuss that in a `conda-build` issue.

Reply:
More discussion in https://github.com/conda-forge/conda-forge-pinning-feedstock/issues/4 and https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/44#issuecomment-383273710

If a single package can be used multiple ways, run_exports and the global pinning are not appropriate.  For boost, there should be header-only uses and linked uses, with only the latter requiring run_exports.  In https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/44#issuecomment-382797245, @isuruf wants to use the global pinnings for them, but leave it up to the recipe author to ignore it where it isn't necessary.

The real solution isn't to do either pin_compatible (local) or pin_run_as_build (global), though either may be interim solutions.  The real solution is to make it so that a single package is not ambiguous for how its run_exports should be set.  This means splitting up or creating metapackages for numpy and libboost that allow us to indicate specific usage that implies linking (and thus binary compatibility concerns/precautions).